modules/hook-commands/hook-commands.c

```diff
@@ -27,30 +27,45 @@
 
 #define HOOK_COMMANDS_PLUGIN "hook-commands"
 
+struct _HookCommandsPlugin
+{
+  LogDriverPlugin super;
+  gchar *startup;
+  gchar *setup;
+  gchar *teardown;
+  gchar *shutdown;
+  gboolean (*saved_init)(LogPipe *s);
+  gboolean (*saved_deinit)(LogPipe *s);
+};
+
 void
-hook_commands_plugin_set_startup(HookCommandsPlugin *self, const gchar *startup)
+hook_commands_plugin_set_startup(LogDriverPlugin *s, const gchar *startup)
 {
+  HookCommandsPlugin *self = (HookCommandsPlugin *)s;
   g_free(self->startup);
   self->startup = g_strdup(startup);
 }
 
 void
-hook_commands_plugin_set_setup(HookCommandsPlugin *self, const gchar *setup)
+hook_commands_plugin_set_setup(LogDriverPlugin *s, const gchar *setup)
 {
+  HookCommandsPlugin *self = (HookCommandsPlugin *)s;
   g_free(self->setup);
   self->setup = g_strdup(setup);
 }
 
 void
-hook_commands_plugin_set_teardown(HookCommandsPlugin *self, const gchar *teardown)
+hook_commands_plugin_set_teardown(LogDriverPlugin *s, const gchar *teardown)
 {
+  HookCommandsPlugin *self = (HookCommandsPlugin *)s;
   g_free(self->teardown);
   self->teardown = g_strdup(teardown);
 }
 
 void
-hook_commands_plugin_set_shutdown(HookCommandsPlugin *self, const gchar *shutdown)
+hook_commands_plugin_set_shutdown(LogDriverPlugin *s, const gchar *shutdown)
 {
+  HookCommandsPlugin *self = (HookCommandsPlugin *)s;
```

Comment:
In the grammar, where it is called, there is no reference to `HookCommandsPlugin`, but only to `LogDriverPlugin`, as the new function should return with the base class.

So either in the grammar the `LogDriverPlugin` should be casted to `HookCommandsPlugin` or in the setters functions.
I would like to keep the grammar clean, so stuff like casting should not be part of grammar if possible.

Why casting multiple time is a disadvantage ?
* The source code gets bigger, but I think it is not a concern in this case.
* Run time it is free(, and even if it would not be it is configuration time.)
* Compile time is probably bigger, but still not a concern I guess.

Reply:
I mean `last_hook_commands` with type `HookCommandsPlugin` was in the grammar. With that, we don't have to cast anything at all.

There was only one cast in the grammar:
```
*instance = (LogDriverPlugin *) last_hook_commands;
```
But this is an "upcast", which is much acceptable in my opinion (and we have only one instead of N).